retrieval/discovery/kubernetes/discovery.go

```diff
@@ -292,18 +297,31 @@ func (kd *Discovery) updateNodesTargetGroup() *config.TargetGroup {
 
 	// Now let's loop through the nodes & add them to the target group with appropriate labels.
 	for nodeName, node := range kd.nodes {
-		nodeAddress, err := nodeHostIP(node)
+		defaultNodeAddress, nodeAddressMap, err := nodeAddresses(node)
 		if err != nil {
 			log.Debugf("Skipping node %s: %s", node.Name, err)
 			continue
 		}
 
-		address := fmt.Sprintf("%s:%d", nodeAddress.String(), kd.Conf.KubeletPort)
+		kubeletPort := int(node.Status.DaemonEndpoints.KubeletEndpoint.Port)
+		if kubeletPort == 0 {
+			kubeletPort = kd.Conf.KubeletPort
+		}
+
+		address := fmt.Sprintf("%s:%d", defaultNodeAddress.String(), kubeletPort)
 
 		t := model.LabelSet{
 			model.AddressLabel:  model.LabelValue(address),
 			model.InstanceLabel: model.LabelValue(nodeName),
 		}
+
+		for addrType, ip := range nodeAddressMap {
+			labelName := strutil.SanitizeLabelName(nodeAddressPrefix + toSnake(string(addrType)))
```

Comment:
So leave labels as `__meta_kubernetes_node_address_InternalIP`, etc? Don't mind either way personally but seems tidier to snake case it.